### PR TITLE
[CRT-1953] Support Pre-encode for videos that have dynamic frame resolutions

### PR DESCRIPTION
**Pull request by MassimoDeLuca**
Following investigations into CRT-1953, trims have problems with Audio / Video sync or duration checks when they contain frames with differing resolutions.
These particular videos have a resolution in the metadata, but individual frames can have different resolutions which is namely a problem with `filter_complex`

Pre-encoding alike how we do for stitches seems to resolve this issue. 
Note this does increase processing time for these particular videos. `permit_dynamic_resolution_pre_encode` is available to allow alchemist to allow this to only happen for trims and not all videos as normal encodes are not affected.

Also modifies stitching to do frame resolution checks to ensure we have the highest resolution amongst the frames incase the metadata is incorrect.

Paired with https://github.com/Vidyard/Alchemist/pull/1240

### CODE REVIEW

File: lib/ffmpeg/movie.rb [hunk 1]
```diff
@@ -264,6 +269,51 @@ def any_streams_contain_audio?
...
+      max_height = height || 0
+      differing_frame_resolutions = false
+      last_dimensions = nil
+
+      unescaped_paths.each do |path|
+        local_movie = Movie.new(path)
+
+        # set max width and height from the movie metadata first
+        max_width = [max_width, local_movie.width].max
+        max_height = [max_height, local_movie.height].max
```
Comment (Kyrluckechuck) [inline lib/ffmpeg/movie.rb:284]: Why set this based on the metadata if we're going to set it based on the actual output from the command below?

File: lib/ffmpeg/transcoder.rb [hunk 2]
```diff
@@ -18,11 +18,22 @@ def self.timeout
       @@timeout
     end
 
+    def requires_pre_encode
+      # Stitches and trims with dynamic resolution require pre-encoding
+      @movie.requires_pre_encode = @movie.paths.size > 1 || (@movie.has_dynamic_resolution && @transcoder_options[:permit_dynamic_resolution_pre_encode])
```
Comment (Kyrluckechuck) [inline lib/ffmpeg/transcoder.rb]: nit: Should this be updated each time it's called, or should it only be updated if it hasn't been set?
```suggestion
      @movie.requires_pre_encode ||= @movie.paths.size > 1 || (@movie.has_dynamic_resolution && @transcoder_options[:permit_dynamic_resolution_pre_encode])
```
If we did go that route, you'd want to update this
```ruby
     @requires_pre_encode = false
```
To this
```ruby     
     @requires_pre_encode = nil
```

**Review by Kyrluckechuck — COMMENTED** [review]
Just a few comments/questions/nits. Overall looks good and appreciate the tests!

File: lib/ffmpeg/movie.rb (hunk 1 above)
Comment (MassimoDeLuca) [reply lib/ffmpeg/movie.rb:284]: This was retained from `calculate_interim_max_dimensions` incase theres a chance the metadata reports a higher resolution than the frames. No preference myself if its kept / removed!

Comment (Kyrluckechuck) [reply lib/ffmpeg/movie.rb:284]: I don't see major harm and it's an incredibly "cheap" transaction, so should be fine

File: lib/ffmpeg/movie.rb [hunk 3]
```diff
@@ -16,7 +16,6 @@ class Movie
     UNSUPPORTED_CODEC_PATTERN = /^Unsupported codec with id (\d+) for input stream (\d+)$/
 
     @has_dynamic_resolution = false
-    @requires_pre_encode = false
```
Comment (Kyrluckechuck) [inline lib/ffmpeg/movie.rb]: nit: It might be nice to keep this still even though we're assigning it's value to `nil` to indicate it _should_ be a variable accessible in the API / explicitly nil until set 🤔

File: lib/ffmpeg/transcoder.rb [hunk 4]
```diff
@@ -105,6 +105,9 @@ def pre_encode_if_necessary
       @movie.unescaped_paths.each_with_index do |path, index|
         audio_map = determine_audio_for_pre_encode(path)
 
+        puts "pre_encode_options: #{pre_encode_options}"
+        puts "max_width: #{max_width}"
+        puts "max_height: #{max_height}"
```
Comment (Kyrluckechuck) [inline lib/ffmpeg/transcoder.rb]: nit: remove logs

File: lib/ffmpeg/transcoder.rb [hunk 5]
```diff
@@ -95,7 +105,13 @@ def pre_encode_if_necessary
       @movie.unescaped_paths.each_with_index do |path, index|
         audio_map = determine_audio_for_pre_encode(path)
 
-        command = "#{@movie.ffmpeg_command} -y -i #{Shellwords.escape(path)} -movflags faststart #{pre_encode_options} -r #{output_frame_rate} -filter_complex \"[0:v]scale=#{max_width}:#{max_height}:force_original_aspect_ratio=decrease,pad=#{max_width}:#{max_height}:(ow-iw)/2:(oh-ih)/2:color=black,setsar=1[Scaled]\" -map \"[Scaled]\" #{audio_map} #{@movie.interim_paths[index]}"
+        # Only re-scale the video if there are multiple inputs. If a single input video contains dynamic resolution it can cause issues with the filter_complex filter
+        complex_filter = ""
+        if @movie.paths.size > 1
```
Comment (Kyrluckechuck) [inline lib/ffmpeg/transcoder.rb:110]: This is no longer using the calculated max_width & max_height -- is that intentional?

Comment (MassimoDeLuca) [reply lib/ffmpeg/transcoder.rb:110]: It still is if this evaluates to true.
It doesn't seem the pre-encode needs it for when this is a single input trim as ffmpeg sizes it to the `-s` argument on its own

Comment (Kyrluckechuck) [reply lib/ffmpeg/transcoder.rb:110]: And when a video contains multiple resolutions, does it correctly not squish/stretch the larger resolution portion?

Comment (MassimoDeLuca) [reply lib/ffmpeg/transcoder.rb:110]: It is correct!
Heres 3 trims stitched together in various orders. (Stitching trim + trim + trim has no issues)
https://staging-aquaman.vidyard.com/organizations/982359/players/ZXFveGaFPaUe7dSW6TsXDe?edit=true
https://staging-aquaman.vidyard.com/organizations/982359/players/E6AwKXKyTpWRU4kVjt4RcV?edit=true
https://staging-aquaman.vidyard.com/organizations/982359/players/7oEWEuuNeq1t6W9C7QKbZJ?edit=true

Heres an additional 4 variations of stitches
https://staging-aquaman.vidyard.com/organizations/982359/players/yqxDfCKFiuDCsWth9zTgP4?edit=true
https://staging-aquaman.vidyard.com/organizations/982359/players/iTNmJPZPTBM8mEdoDbWTMu?edit=true
https://staging-aquaman.vidyard.com/organizations/982359/players/RtmZKzS7UGQGyzsJMU6PSk?edit=true
https://staging-aquaman.vidyard.com/organizations/982359/players/HkDmFS3dvJBYpDYPer6vuv?edit=true

File: lib/ffmpeg/transcoder.rb [hunk 6]
```diff
@@ -48,7 +59,6 @@ def initialize(movie, output_file, options = EncodingOptions.new, transcoder_opt
         raise ArgumentError, "Unknown options format '#{options.class}', should be either EncodingOptions, Hash or String."
       end
 
-      @transcoder_options = transcoder_options
```
Comment (Kyrluckechuck) [inline lib/ffmpeg/transcoder.rb:51]: Why was this removed?

Comment (MassimoDeLuca) [reply lib/ffmpeg/transcoder.rb:51]: This was moved up a few lines so it was accessible to `requires_pre_encode`

**Review by Kyrluckechuck — APPROVED** [review]
🤞
